### PR TITLE
Support server-side streaming of a unary RPC

### PR DESCRIPTION
**Pull request by vjpai**
This mode of execution behaves like a unary RPC on the client-side but like a single-shot streaming RPC on the server-side (proper function means exactly 1 read and 1 write). The purpose is to allow the server to delay the actual read of the RPC request until it is ready to do so (e.g., for memory requirements).

There is some testing in place now as part of hybrid_end2end_test

Cc: @shaladdle

### CODE REVIEW

**Comment by shaladdle** [comment]
Thanks for putting this together!

Two main concerns:

1. We'd really like to be able to have one base class (ServiceName::FCService or something?) which has all unary methods as FCUnary, and also has streaming methods. With the current PR, we'd have to make ~15 classes in order to flow control every method. That by itself seems kind of messy, but instantiation of our service impl would be messy as well, since we'd have to wrap 15 times.
2. Wouldn't it make sense for FCUnary to inherit from ServerReaderWriter? We envision writing some helper code that works for both streaming and unary, and if it at least inherited from ServerReaderWriter interface, that would make it easier.

I'll try patching this and start experimenting with it.

File: include/grpc++/impl/codegen/fc_unary.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+/// the message.
+/// A key difference relative to streaming: an FCUnary must have exactly 1 Read
+/// and exactly 1 Write, in that order, to function correctly.
+/// Otherwise, the RPC is in error.
+template <class RequestType, class ResponseType>
+  class FCUnary GRPC_FINAL {
+ public:
+ FCUnary(Call* call, ServerContext* ctx): call_(call), ctx_(ctx), read_done_(false), write_done_(false) {}
+  ~FCUnary() {}
+  uint32_t NextMessageSize() {return call_->max_message_size();}
```
Comment (shaladdle) [inline include/grpc++/impl/codegen/fc_unary.h]: What happens if the other side drops the connection? Maybe this signature (and the Reader signature) should be

```
bool NextMessageSize(uint32_t* size)
```

?

File: include/grpc++/impl/codegen/sync_stream.h [hunk 2]
```diff
@@ -70,6 +70,9 @@ class ReaderInterface {
  public:
   virtual ~ReaderInterface() {}
 
+  /// Upper bound on the next message size available for reading on this stream
+  virtual uint32_t NextMessageSize() = 0;
```
Comment (shaladdle) [inline include/grpc++/impl/codegen/sync_stream.h]:
```
bool NextMessageSize(uint32_t* message_size)
```

?

**Comment by vjpai** [comment]
It's been a while, but I now agree with you wrt this inheriting from ServerReaderWriter. It is a special-case of streaming that only allows exactly 1 Read and exactly 1 Write in that order, but that still merits considering it as though it were streaming. Will also get that other typedef in soon.

**Comment by vjpai** [comment]
The last push was to just to get this PR unconflicted; will make requested changes next.

**Comment by vjpai** [comment]
1.  `ServerReaderWriter` and `FCUnary` now inherit from a common `ServerReaderWriterInterface`

2.  `FCUnaryService` is now a typedef for a service with all unary methods converted to FCUnary, much like `AsyncService` is a typedef for a service with all methods converted to Async.

**Comment by vjpai** [comment]
@a11r notes that NextMessageSize only refers to message size as it appears on the network but that actual parsed proto size is likely to be much larger.

**Comment by vjpai** [comment]
@ctiller : this is ready for review. Thanks!

File: include/grpc++/impl/codegen/status_code_enum.h [hunk 3]
```diff
@@ -143,6 +143,11 @@ enum StatusCode {
   /// Unrecoverable data loss or corruption.
   DATA_LOSS = 15,
 
+  // Service was improperly implemented, violated a gRPC API requirement
```
Comment (ctiller) [inline include/grpc++/impl/codegen/status_code_enum.h]: This error space is not owned by us, and would need to go through some... process to be added (will chat offline)

Comment (vjpai) [inline include/grpc++/impl/codegen/status_code_enum.h]: Ok, I've now changed this to use INTERNAL since this is also an invariant failure.

**Comment by ctiller** [comment]
LGTM

**Comment by vjpai** [comment]
Cc: @y-zeng FYI

**Comment by vjpai** [comment]
Although I think the remaining test failures are unrelated, I want them to look better before tagging as ready to merge. Rerunning the tests.

**Comment by vjpai** [comment]
Current Mac-Opt builds are failing across PRs, so those failures are not from this PR. Suggest tagging ready-to-merge, but can hold off if desired.

**Pull request merged**